src/server/scripts/Spells/spell_priest.cpp

```diff
@@ -2846,6 +2847,52 @@ class spell_pri_trail_of_light : public AuraScript
     std::queue<ObjectGuid> _healQueue;
 };
 
+// 390693 - Train of Thought
+// Called by Flash Heal, Renew, Smite
+class spell_pri_train_of_thought : public AuraScript
+{
+    bool Validate(SpellInfo const* /*spellInfo*/) override
+    {
+        return ValidateSpellInfo
+        ({
+            SPELL_PRIEST_FLASH_HEAL,
+            SPELL_PRIEST_RENEW,
+            SPELL_PRIEST_SMITE,
+            SPELL_PRIEST_POWER_WORD_SHIELD,
+            SPELL_PRIEST_PENANCE
+        });
+    }
+
+    void HandleProc(ProcEventInfo& eventInfo)
+    {
+        Unit* caster = GetTarget();
+        if (!caster)
+            return;
+
+        if (eventInfo.GetSpellInfo()->Id == SPELL_PRIEST_SMITE)
+        {
+            SpellInfo const* penance = sSpellMgr->GetSpellInfo(SPELL_PRIEST_PENANCE, GetCastDifficulty());
+            AuraEffect const* aurEff = caster->GetAuraEffect(SPELL_PRIEST_TRAIN_OF_THOUGHT, EFFECT_1);
```

Comment:
you need to validate this effect, OnProc hook does not implicitly validate effects regardless of not being scripting a specific effect